sw/device/lib/ujson/crc32.h

```diff
@@ -0,0 +1,49 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_UJSON_CRC32_H_
+#define OPENTITAN_SW_DEVICE_LIB_UJSON_CRC32_H_
+#include "sw/device/lib/ujson/ujson_derive.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// clang-format off
+/////////////////////////////////////////////////////////////////////////////
+// Automatic generation of structs with serialize/deserialize functions:
+//
+// The following creates a `struct Crc32`:
+// typedef struct Crc32 {
+//     uint32_t crc32;
+// } crc32_t;
+// status_t ujson_serialize_crc32_t(ujson_t *context, const crc32_t *self);
+// status_t ujson_deserialize_crc32_t(ujson_t *context, crc32_t *self);
+#define STRUCT_CRC32(field, string) \
+    field(crc32, uint32_t)
+UJSON_SERDE_STRUCT(Crc32, crc32_t, STRUCT_CRC32);
```

Comment:
How do you imagine integrating the CRC32 into the console based message exchanging protocol?

There is certainly nothing wrong with encoding the CRC32 this way, but I've always thought of the crc32 as sort-of an out-of-band optional message for validating the previous JSON blob.  On the rust client side of the protocol, the receiver is looking for a [newline-terminated regex](https://github.com/lowRISC/opentitan/blob/master/sw/host/opentitanlib/src/test_utils/rpc.rs#L43-L44).

If this were transmitted as json, then some amount of parsing is required to separate the first json message from the crc message.  If, on the other hand, we were to encode this as some out-of-band text, the regex could easily match that and we could validate the CRC over the json message before even attempting any parsing or decoding:

```rust
    exit_success: Some(Regex::new(r"RESP_OK:(.*) (CRC:(\d+))?\r")),
    exit_failure: Some(Regex::new(r"RESP_ERR:(.*) (CRC:(\d+))?\r")),
```

Then later, after the console captures the regex, we can do something like the following pseudo-code:
```rust
let msg = captures.get(1)?;
let crc = u32::from_str(captures.get(2)?)?;
check_crc(&msg, crc).map_err(CrcError)?;
let decoded_message = serde_json::from_str(&msg)?;
return Ok(decoded_message);
```

We can, of course, do this with the CRC encoded as json, but we then either need additional parsing or a more complex regex to match the CRC portion of the message.

Reply:
Yep, I guess `crc32_t` is somewhat superfluous.